Lambda School Data Science

*Unit 1, Sprint 1, Module 3*

---

# Join and Reshape datasets

Objectives
- concatenate data with pandas
- merge data with pandas
-  understand tidy data formatting
-  melt and pivot data with pandas

Links
- [Pandas Cheat Sheet](https://github.com/pandas-dev/pandas/blob/master/doc/cheatsheet/Pandas_Cheat_Sheet.pdf)
- [Tidy Data](https://en.wikipedia.org/wiki/Tidy_data)
  - Combine Data Sets: Standard Joins
  - Tidy Data
  - Reshaping Data
- Python Data Science Handbook
  - [Chapter 3.6](https://jakevdp.github.io/PythonDataScienceHandbook/03.06-concat-and-append.html), Combining Datasets: Concat and Append
  - [Chapter 3.7](https://jakevdp.github.io/PythonDataScienceHandbook/03.07-merge-and-join.html), Combining Datasets: Merge and Join
  - [Chapter 3.8](https://jakevdp.github.io/PythonDataScienceHandbook/03.08-aggregation-and-grouping.html), Aggregation and Grouping
  - [Chapter 3.9](https://jakevdp.github.io/PythonDataScienceHandbook/03.09-pivot-tables.html), Pivot Tables
  
Reference
- Pandas Documentation: [Reshaping and Pivot Tables](https://pandas.pydata.org/pandas-docs/stable/reshaping.html)
- Modern Pandas, Part 5: [Tidy Data](https://tomaugspurger.github.io/modern-5-tidy.html)

In [12]:
!wget https://s3.amazonaws.com/instacart-datasets/instacart_online_grocery_shopping_2017_05_01.tar.gz

--2020-03-18 20:49:15--  https://s3.amazonaws.com/instacart-datasets/instacart_online_grocery_shopping_2017_05_01.tar.gz
Resolving s3.amazonaws.com (s3.amazonaws.com)... 52.216.241.30
Connecting to s3.amazonaws.com (s3.amazonaws.com)|52.216.241.30|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 205548478 (196M) [application/x-gzip]
Saving to: ‘instacart_online_grocery_shopping_2017_05_01.tar.gz’

instacart_online_gr 100%[===================>] 196.03M  42.4MB/s    in 4.9s    

2020-03-18 20:49:20 (40.0 MB/s) - ‘instacart_online_grocery_shopping_2017_05_01.tar.gz’ saved [205548478/205548478]



In [13]:
!tar --gunzip --extract --verbose --file=instacart_online_grocery_shopping_2017_05_01.tar.gz

instacart_2017_05_01/
instacart_2017_05_01/._aisles.csv
instacart_2017_05_01/aisles.csv
instacart_2017_05_01/._departments.csv
instacart_2017_05_01/departments.csv
instacart_2017_05_01/._order_products__prior.csv
instacart_2017_05_01/order_products__prior.csv
instacart_2017_05_01/._order_products__train.csv
instacart_2017_05_01/order_products__train.csv
instacart_2017_05_01/._orders.csv
instacart_2017_05_01/orders.csv
instacart_2017_05_01/._products.csv
instacart_2017_05_01/products.csv


In [14]:
%cd instacart_2017_05_01

/content/instacart_2017_05_01


In [15]:
!ls -lh *.csv

-rw-r--r-- 1 502 staff 2.6K May  2  2017 aisles.csv
-rw-r--r-- 1 502 staff  270 May  2  2017 departments.csv
-rw-r--r-- 1 502 staff 551M May  2  2017 order_products__prior.csv
-rw-r--r-- 1 502 staff  24M May  2  2017 order_products__train.csv
-rw-r--r-- 1 502 staff 104M May  2  2017 orders.csv
-rw-r--r-- 1 502 staff 2.1M May  2  2017 products.csv


# Assignment

## Join Data Practice

These are the top 10 most frequently ordered products. How many times was each ordered? 

1. Banana
2. Bag of Organic Bananas
3. Organic Strawberries
4. Organic Baby Spinach 
5. Organic Hass Avocado
6. Organic Avocado
7. Large Lemon 
8. Strawberries
9. Limes 
10. Organic Whole Milk

First, write down which columns you need and which dataframes have them.

Next, merge these into a single dataframe.

Then, use pandas functions from the previous lesson to get the counts of the top 10 most frequently ordered products.

In [0]:
##### YOUR CODE HERE #####
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline


In [8]:
aisles = pd.read_csv('aisles.csv')
print(aisles.shape)
aisles.head()

(134, 2)


aisle_id                       aisle
0         1       prepared soups salads
1         2           specialty cheeses
2         3         energy granola bars
3         4               instant foods
4         5  marinades meat preparation

In [9]:
departments = pd.read_csv('departments.csv')
print(departments.shape)
departments.head()

(21, 2)


department_id department
0              1     frozen
1              2      other
2              3     bakery
3              4    produce
4              5    alcohol

In [10]:
# Orders
orders = pd.read_csv('orders.csv')
print(orders.shape)
orders.head()

(3421083, 7)


order_id  user_id  ... order_hour_of_day  days_since_prior_order
0   2539329        1  ...                 8                     NaN
1   2398795        1  ...                 7                    15.0
2    473747        1  ...                12                    21.0
3   2254736        1  ...                 7                    29.0
4    431534        1  ...                15                    28.0

[5 rows x 7 columns]

In [11]:
# Products
products = pd.read_csv('products.csv')
print(products.shape)
products.head()

(49688, 4)


product_id  ... department_id
0           1  ...            19
1           2  ...            13
2           3  ...             7
3           4  ...             1
4           5  ...            13

[5 rows x 4 columns]

In [12]:
orders = pd.read_csv('orders.csv')
orders.head(10)

order_id  user_id  ... order_hour_of_day  days_since_prior_order
0   2539329        1  ...                 8                     NaN
1   2398795        1  ...                 7                    15.0
2    473747        1  ...                12                    21.0
3   2254736        1  ...                 7                    29.0
4    431534        1  ...                15                    28.0
5   3367565        1  ...                 7                    19.0
6    550135        1  ...                 9                    20.0
7   3108588        1  ...                14                    14.0
8   2295261        1  ...                16                     0.0
9   2550362        1  ...                 8                    30.0

[10 rows x 7 columns]

In [0]:
cols=['user_id', 'order_id', 'order_number', 'order_dow', 'order_hour_of_day']
orders=orders[cols]

In [36]:
condition = (orders['user_id']==1) & (orders['order_number']<=2)
condition

0           True
1           True
2          False
3          False
4          False
           ...  
3421078    False
3421079    False
3421080    False
3421081    False
3421082    False
Length: 3421083, dtype: bool

In [0]:
orders[condition]

In [29]:
print(orders[condition].shape)
orders[condition]

(2, 5)


user_id  order_id  order_number  order_dow  order_hour_of_day
0        1   2539329             1          2                  8
1        1   2398795             2          3                  7

In [19]:
condition = (order_products['order_id']==2539329) | (order_products['order_id']==2538795)
condition

0         True
1        False
2        False
3        False
4        False
         ...  
49683    False
49684    False
49685    False
49686    False
49687    False
Name: order_id, Length: 3470771, dtype: bool

In [23]:
columns = ['order_id', 'add_to_cart_order', 'product_name']
columns

['order_id', 'add_to_cart_order', 'product_name']

In [38]:
order_products_subset = order_products.loc[condition, columns]
order_products_subset

/usr/local/lib/python3.6/dist-packages/pandas/core/indexing.py:1418: FutureWarning: 
Passing list-likes to .loc or [] with any missing label will raise
KeyError in the future, you can use .reindex() as an alternative.

See the documentation here:
https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#deprecate-loc-reindex-listlike
  return self._getitem_tuple(key)


order_id  add_to_cart_order                product_name
0  2539329.0                NaN                         NaN
1  2398795.0                NaN                         NaN
0        NaN                NaN  Chocolate Sandwich Cookies
1        NaN                NaN            All-Seasons Salt

In [39]:
merged = pd.merge(order_products, order_products_subset, on='order_id', how='inner')
merged

aisle_id  department_id  ...  add_to_cart_order              product_name_y
0           NaN            NaN  ...                NaN                         NaN
1           NaN            NaN  ...                NaN                         NaN
2          61.0           19.0  ...                NaN  Chocolate Sandwich Cookies
3          61.0           19.0  ...                NaN            All-Seasons Salt
4         104.0           13.0  ...                NaN  Chocolate Sandwich Cookies
...         ...            ...  ...                ...                         ...
99373     112.0            3.0  ...                NaN            All-Seasons Salt
99374      41.0            8.0  ...                NaN  Chocolate Sandwich Cookies
99375      41.0            8.0  ...                NaN            All-Seasons Salt
99376      73.0           11.0  ...                NaN  Chocolate Sandwich Cookies
99377      73.0           11.0  ...                NaN            All-Seasons Salt

[99378 rows x 11 columns]

In [41]:
final = pd.merge(merged, products[['product_id', 'product_name']], how='inner', on='product_id')
final

aisle_id  ...                                product_name
0          61.0  ...                  Chocolate Sandwich Cookies
1          61.0  ...                  Chocolate Sandwich Cookies
2         104.0  ...                            All-Seasons Salt
3         104.0  ...                            All-Seasons Salt
4          94.0  ...        Robust Golden Unsweetened Oolong Tea
...         ...  ...                                         ...
99371     112.0  ...                            Artisan Baguette
99372      41.0  ...  Smartblend Healthy Metabolism Dry Cat Food
99373      41.0  ...  Smartblend Healthy Metabolism Dry Cat Food
99374      73.0  ...                      Fresh Foaming Cleanser
99375      73.0  ...                      Fresh Foaming Cleanser

[99376 rows x 12 columns]

In [18]:
products_df = pd.read_csv('products.csv')
products_df

product_id  ... department_id
0               1  ...            19
1               2  ...            13
2               3  ...             7
3               4  ...             1
4               5  ...            13
...           ...  ...           ...
49683       49684  ...             5
49684       49685  ...             1
49685       49686  ...             3
49686       49687  ...             8
49687       49688  ...            11

[49688 rows x 4 columns]

In [20]:
pd.set_option('display.max_rows', 100)
products_df.isnull().sum().sort_values(ascending=False)

department_id    0
aisle_id         0
product_name     0
product_id       0
dtype: int64

In [29]:
products_df['product_name'].value_counts().head(10)

Clif Bar Nut Butter Filed Chocolate Hazelnut Butter 4 Ct 7.04 Oz    1
Protein Shake, Organic, Banana Creme                                1
Teriyaki Baste & Glaze                                              1
Dulce de Leche Caramel Ice Cream                                    1
Smoked Salt Caramels in Milk Chocolate                              1
Foods Gourmet Dry Salami Black Pepper                               1
Mini Toasts                                                         1
Face for Everyday Moisturize                                        1
Classic Basil Pesto Sauce                                           1
Potato & Ricotta Gnocchi                                            1
Name: product_name, dtype: int64

## Reshape Data Section

- Replicate the lesson code. Remember, if you haven't followed along typing out what we did during lecture, do that now to get more repetition with the syntax.
- Take table 2 (the transpose of table 1) and practice changing it into Tidy format and back again. You should not use the transpose operation anywhere in this code:
 - Table 2 --> Tidy
 - Tidy --> Table 2
- Load seaborn's `flights` dataset by running the cell below. Then create a pivot table showing the number of passengers by month and year. Use year for the index and month for the columns. You've done it right if you get 112 passengers for January 1949 and 432 passengers for December 1960.

In [30]:
import pandas as pd
import numpy as np

table1 = pd.DataFrame(
    [[np.nan, 2],
     [16,    11], 
     [3,      1]],
    index=['John Smith', 'Jane Doe', 'Mary Johnson'], 
    columns=['treatmenta', 'treatmentb'])

table2 = table1.T

table2

John Smith  Jane Doe  Mary Johnson
treatmenta         NaN      16.0           3.0
treatmentb         2.0      11.0           1.0

In [32]:
# table2 -> tidy
table2.columns.tolist()

['John Smith', 'Jane Doe', 'Mary Johnson']

In [33]:
table2.index.tolist()

['treatmenta', 'treatmentb']

In [34]:
table2

John Smith  Jane Doe  Mary Johnson
treatmenta         NaN      16.0           3.0
treatmentb         2.0      11.0           1.0

In [36]:
table2 = table2.reset_index()
table2

level_0       index  John Smith  Jane Doe  Mary Johnson
0        0  treatmenta         NaN      16.0           3.0
1        1  treatmentb         2.0      11.0           1.0

In [40]:
tidy = table2.melt(id_vars= 'index',
                       value_vars = ['John Smith', 'Jane Doe', 'Mary Johnson'])
tidy

index      variable  value
0  treatmenta    John Smith    NaN
1  treatmentb    John Smith    2.0
2  treatmenta      Jane Doe   16.0
3  treatmentb      Jane Doe   11.0
4  treatmenta  Mary Johnson    3.0
5  treatmentb  Mary Johnson    1.0

In [42]:
# tidy -> table2
tidy['index'] = tidy['index'].str.replace('treatment', '')
tidy

index      variable  value
0     a    John Smith    NaN
1     b    John Smith    2.0
2     a      Jane Doe   16.0
3     b      Jane Doe   11.0
4     a  Mary Johnson    3.0
5     b  Mary Johnson    1.0

In [0]:
import seaborn as sns

In [0]:
flights = sns.load_dataset('flights')

In [46]:
# Flights Pivot Table
flights

year      month  passengers
0    1949    January         112
1    1949   February         118
2    1949      March         132
3    1949      April         129
4    1949        May         121
..    ...        ...         ...
139  1960     August         606
140  1960  September         508
141  1960    October         461
142  1960   November         390
143  1960   December         432

[144 rows x 3 columns]

In [49]:
flights1 = tidy.pivot_table
flights1

<bound method DataFrame.pivot_table of   index      variable  value
0     a    John Smith    NaN
1     b    John Smith    2.0
2     a      Jane Doe   16.0
3     b      Jane Doe   11.0
4     a  Mary Johnson    3.0
5     b  Mary Johnson    1.0>

## Join Data Stretch Challenge

The [Instacart blog post](https://tech.instacart.com/3-million-instacart-orders-open-sourced-d40d29ead6f2) has a visualization of "**Popular products** purchased earliest in the day (green) and latest in the day (red)." 

The post says,

> "We can also see the time of day that users purchase specific products.

> Healthier snacks and staples tend to be purchased earlier in the day, whereas ice cream (especially Half Baked and The Tonight Dough) are far more popular when customers are ordering in the evening.

> **In fact, of the top 25 latest ordered products, the first 24 are ice cream! The last one, of course, is a frozen pizza.**"

Your challenge is to reproduce the list of the top 25 latest ordered popular products.

We'll define "popular products" as products with more than 2,900 orders.



In [0]:
##### YOUR CODE HERE #####

## Reshape Data Stretch Challenge

_Try whatever sounds most interesting to you!_

- Replicate more of Instacart's visualization showing "Hour of Day Ordered" vs "Percent of Orders by Product"
- Replicate parts of the other visualization from [Instacart's blog post](https://tech.instacart.com/3-million-instacart-orders-open-sourced-d40d29ead6f2), showing "Number of Purchases" vs "Percent Reorder Purchases"
- Get the most recent order for each user in Instacart's dataset. This is a useful baseline when [predicting a user's next order](https://www.kaggle.com/c/instacart-market-basket-analysis)
- Replicate parts of the blog post linked at the top of this notebook: [Modern Pandas, Part 5: Tidy Data](https://tomaugspurger.github.io/modern-5-tidy.html)

In [0]:
##### YOUR CODE HERE #####